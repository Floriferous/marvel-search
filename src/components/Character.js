import React from 'react';
import PropTypes from 'prop-types';

import CharacterImage from './CharacterImage';
import CharacterFooter from './CharacterFooter';

const Character = ({ name, thumbnail }) => (
  <li className="character">
    <div className="content">
      <CharacterImage name={name} {...thumbnail} />
      <CharacterFooter name={name} />
    </div>
    <p className="helper">Click to bookmark</p>
  </li>
);

Character.propTypes = {};

export default Character;
