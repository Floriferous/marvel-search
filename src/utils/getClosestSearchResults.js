import { createSearchKey } from '../actions/search';

/**
 * getClosestSearchResults - Searches for the closest search results
 * currently in memory. It iterates on a search string, and removes one
 * character at a time until it finds an old search result to display.
 *
 * If the first run passes, it simply returns the exact search results.
 *
 * An isLoading flag is also returned, to indicate whether old or exact
 * results are currently being returned
 *
 * @param {Object} searchResults Object with all previous search results
 * @param {String} search        current search string
 * @param {String} pagination    current pagination
 *
 * @return {Object}
 */
const getClosestSearchResults = (searchResults, search, pagination) => {
  let broaderResults;
  let isExactSearch;

  // Return immediately if search length equals 1 or pagination is larger than 0
  if (search.length === 1 || pagination > 0) {
    return {
      data: searchResults[createSearchKey(search, pagination)] || {},
      isLoading: false, // taken care of by CharacterList
    };
  }

  for (let i = 0; i < search.length; i += 1) {
    const broaderSearch = search.slice(0, search.length - i);
    broaderResults = searchResults[createSearchKey(broaderSearch, pagination)];
    if (broaderResults) {
      isExactSearch = i === 0;
      break;
    }
  }
  return { data: broaderResults || {}, isLoading: !isExactSearch };
};

export default getClosestSearchResults;
