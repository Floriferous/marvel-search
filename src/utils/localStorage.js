export const storageAvailable = () => {
  try {
    const storage = localStorage;
    const x = '__storage_test__';
    storage.setItem(x, x);
    storage.removeItem(x);
    return true;
  } catch (e) {
    return false;
  }
};

export const saveValueAtKey = (key, value) => {
  if (storageAvailable()) {
    localStorage.setItem(key, JSON.stringify(value));
    return true;
  }

  return false;
};

export const getValueAtKey = (key) => {
  if (storageAvailable()) {
    return JSON.parse(localStorage.getItem(key));
  }

  return false;
};
