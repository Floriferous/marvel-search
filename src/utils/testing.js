import React from 'react';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({ adapter: new Adapter() });

export const testComponent = (Component, props) => {
  if (!testComponent.mountedLockScreen) {
    testComponent.mountedLockScreen = Enzyme.shallow(<Component {...props} />);
  }
  return testComponent.mountedLockScreen;
};

testComponent.reset = () => {
  testComponent.mountedLockScreen = undefined;
};

export const { shallow, mount, render } = Enzyme;
export default Enzyme;
