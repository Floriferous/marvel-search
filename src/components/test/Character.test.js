import React from 'react';
import renderer from 'react-test-renderer';

import { testComponent } from '../../utils/testing';
import Character from '../Character';

describe('Character', () => {
  let props;
  const component = () => testComponent(Character, props);

  beforeEach(() => {
    props = {
      character: {
        isBookmarked: false,
        name: 'superman',
        thumbnail: { test: 'test', test2: 'test2' },
      },
      toggleBookmark: jest.fn(),
    };
    testComponent.reset();
  });

  it('matches its snapshot', () => {
    const tree = renderer.create(<Character {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders a list item with class character', () => {
    expect(component().find('li').length).toBe(1);
    expect(component()
      .find('li')
      .hasClass('character')).toBe(true);
  });

  describe('the list item', () => {
    let listItem;
    beforeEach(() => {
      listItem = component()
        .find('li')
        .first()
        .shallow();
    });

    it('renders a div with class content, and not bookmarked', () => {
      expect(listItem.find('div').length).toBe(1);
      expect(listItem.find('div').hasClass('content')).toBe(true);
      expect(listItem.find('div').hasClass('bookmarked')).toBe(false);
    });

    it('renders a CharacterBookmarkHelper with prop isBookmarked', () => {
      expect(listItem.find('CharacterBookmarkHelper').length).toBe(1);
    });

    describe('the content div', () => {
      let div;
      beforeEach(() => {
        div = listItem
          .find('div')
          .first()
          .shallow();
      });

      it('contains the CharacterImage', () => {
        expect(div.find('CharacterImage').length).toBe(1);
      });

      it('contains the CharacterFooter', () => {
        expect(div.find('CharacterFooter').length).toBe(1);
      });
    });
  });

  it('calls toggleBookmark when the div is clicked with character as argument', () => {
    component()
      .find('.content')
      .simulate('click');
    expect(props.toggleBookmark).toHaveBeenCalledWith(props.character);
  });

  it('passes isBookmarked to the CharacterBookmarkHelper', () => {
    expect(component()
      .find('CharacterBookmarkHelper')
      .props().isBookmarked).toBe(props.character.isBookmarked);
  });

  it('passes name to CharacterImage', () => {
    expect(component()
      .find('CharacterImage')
      .props().name).toBe(props.character.name);
  });

  it('passes name to CharacterFooter', () => {
    expect(component()
      .find('CharacterFooter')
      .props().name).toBe(props.character.name);
  });

  it('passes the entire thumbnail to CharacterImage', () => {
    Object.keys(props.character.thumbnail).forEach((key) => {
      expect(component()
        .find('CharacterImage')
        .props()[key]).toBe(props.character.thumbnail[key]);
    });
  });
});
