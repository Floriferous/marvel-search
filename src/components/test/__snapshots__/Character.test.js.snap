// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Character matches its snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Character
    character={
        Object {
            "isBookmarked": false,
            "name": "superman",
            "thumbnail": Object {
              "test": "test",
              "test2": "test2",
            },
          }
    }
    toggleBookmark={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": null,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="content"
          onClick={[Function]}
>
          <CharacterImage
                    name="superman"
                    test="test"
                    test2="test2"
          />
          <CharacterFooter
                    name="superman"
          />
</div>,
        <CharacterBookmarkHelper
          isBookmarked={false}
/>,
      ],
      "className": "character",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": null,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <CharacterImage
              name="superman"
              test="test"
              test2="test2"
/>,
            <CharacterFooter
              name="superman"
/>,
          ],
          "className": "content",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": null,
            "nodeType": "function",
            "props": Object {
              "name": "superman",
              "test": "test",
              "test2": "test2",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": null,
            "nodeType": "function",
            "props": Object {
              "name": "superman",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": null,
        "nodeType": "function",
        "props": Object {
          "isBookmarked": false,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "li",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": null,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="content"
            onClick={[Function]}
>
            <CharacterImage
                        name="superman"
                        test="test"
                        test2="test2"
            />
            <CharacterFooter
                        name="superman"
            />
</div>,
          <CharacterBookmarkHelper
            isBookmarked={false}
/>,
        ],
        "className": "character",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <CharacterImage
                name="superman"
                test="test"
                test2="test2"
/>,
              <CharacterFooter
                name="superman"
/>,
            ],
            "className": "content",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": null,
              "nodeType": "function",
              "props": Object {
                "name": "superman",
                "test": "test",
                "test2": "test2",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": null,
              "nodeType": "function",
              "props": Object {
                "name": "superman",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "function",
          "props": Object {
            "isBookmarked": false,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "li",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
